modules/react/text-input/lib/InputGroup.tsx

```diff
@@ -0,0 +1,137 @@
+import React from 'react';
+
+import {createComponent, ExtractProps, useIsRTL} from '@workday/canvas-kit-react/common';
+import {Flex} from '@workday/canvas-kit-react/layout';
+
+import {TextInput} from './TextInput';
+
+export const InputGroupStart = createComponent('div')({
+  Component(elemProps: ExtractProps<typeof Flex>, ref, Element) {
+    return (
+      <Flex
+        ref={ref}
+        as={Element}
+        position="absolute"
+        alignItems="center"
+        justifyContent="center"
+        height={40}
+        width={40}
+        {...elemProps}
+      />
+    );
+  },
+});
+
+export const InputGroupEnd = createComponent('div')({
+  Component(elemProps: ExtractProps<typeof Flex>, ref, Element) {
+    return (
+      <Flex
+        ref={ref}
+        as={Element}
+        position="absolute"
+        alignItems="center"
+        justifyContent="center"
+        height={40}
+        width={40}
+        {...elemProps}
+      />
+    );
+  },
+});
+
+export const InputGroupInput = createComponent(TextInput)({
+  Component(elemProps: ExtractProps<typeof Flex>, ref, Element) {
+    return <Flex ref={ref} as={Element} width="100%" {...elemProps} />;
+  },
+});
+
+/**
+ * An `InputGroup` is an Input with optional start and end elements. The start and end elements are
+ * usually icons or icon buttons. The `InputGroup` will add padding to the input so the
+ * icons/buttons display correctly. This component uses `React.Children.map` and
+ * `React.cloneElement` from the [React.Children](https://react.dev/reference/react/Children) API.
+ * This means all children must be `InputGroup.*` components. Any other direct children will cause
+ * issues. You can add different elements/components inside the
+ * {@link InputGroupStart InputGroup.Start} and {@link InputGroupEnd InputGroup.End} subcomponents.
+ *
+ * ```tsx
+ * <InputGroup>
+ *   <InputGroup.Start as={SystemIcon} pointerEvents="none" icon={searchIcon} />
+ *   <InputGroup.Input />
+ *   <InputGroup.End>
+ *     <TertiaryButton tabIndex={-1} icon={xIcon} size="small" />
+ *   </InputGroup.End>
+ * </InputGroup>
+ * ```
+ */
+export const InputGroup = createComponent('div')({
+  displayName: 'InputGroup',
+  Component({children, ...elemProps}: ExtractProps<typeof Flex>, ref, Element) {
+    const isRTL = useIsRTL();
+    let paddingInlineStart: number;
+    let paddingInlineEnd: number;
+
+    React.Children.forEach(children, child => {
+      if (React.isValidElement<any>(child) && child.type === InputGroupStart) {
+        paddingInlineStart = (paddingInlineStart || 0) + (child.props.width || 40);
```

Comment:
Do multiple layers of `calc` work?

Reply:
`paddingInlineStart` is a number. I'll try out these changes locally to make sure everything works.

> Should we set the default for the width prop to space.xl here? Would that save you a lot of || logic?

No, because `child.props` is what's being passed to the child, not the default props of the child, which is why the `|| 40` is repeated here.

Reply:
Multiple layers of `calc` work: `calc(calc(0px + 40px) + 40px);`